### DISCUSSION TITLE
How to add multiple site url's for authentication ?

### DISCUSSION THREAD

**Discussion by VaibhavAcharya** [supabase/supabase#600] [Questions]
I uploaded one of my project vercel.
I ran into a problem where it was authenticating me for `http://localhost:3000` which of course did not work.
So I changed site url of authentication settings to my vercel assigned url.
But now I cannot test my application locally because it does not log me in.

Is there a way to assign multiple url's in authentication settings or any other way to tackle this problem  ?

Btw thank you for this beautiful product.

**Comment by AndreaBarghigiani** [comment]
Hi @VaibhavAcharya, 
I run into the same issue and my solution has been to create a new app inside Supabase and assign to it the different URL. Now I am using one supabase for local development and one for production.

I think that this is something that you have to do anyway because, for example, I am handling the login via BitBucket oAuth and even for it I had to create two different app because the redirect was different.

This brings a little problem, mostly I think it's because I am not that good with PostgreSQL, the data will not be cloned from one database to the other but more importantly IMO is the fact that if you change schema you have to make same changes in both. Hope that if someone that's more proficent with PostgreSQL is reading this can jump and help us 😊

**Comment by kiwicopple** [comment]
We're working on this now :) 

https://github.com/supabase/gotrue/issues/63

**Reply by ProjectMoon** [reply]
So it looks like this has been merged. Does the dashboard still need to be updated? It still seems to accept only single websites, and putting in a comma-separated list predictably leads to a broken redirect.

**Reply by jmn** [reply]
I'm wondering what the status is on this as it is not feasible to change the OAuth redirect URL to localhost when testing, nor is keeping cloned projects. As mentioned [here](https://github.com/supabase/supabase/discussions/1514#discussioncomment-739869), firebase handles redirect to localhost if the auth call comes from localhost.

**Reply by Nick-Mazuk** [reply]
I haven't tried it yet, but it looks like the dashboard now has a field named "Additional Redirect Urls" that looks like it may solve this.

Right next to the "Site Url" under Dashboard > Authentication > Settings.

**Reply by jmn** [reply]
Thanks @Nick-Mazuk  but I have seen the field "Additional redirect urls" and I've tried adding e.g. http://localhost:3000 there but supabase will still redirect me to the URL in "Site URL" when I OAuth with GitHub. Perhaps I'm missing some piece of configuration?

**Reply by ptcampbell** [reply]
@jmn did you find a way to do what you want?

**Reply by jmn** [reply]
@ptcampbell Unfortunately not. Hoping for a reply from the devs.

**Reply by ajoslin103** [reply]
I too am having this issue -- I can either redirect to my localhost:8910 or to my Netlify deployment, but not both...

I suspect that the different projects within Supabase are the eventual way to go as I will not want my prod deployment to share my dev database...

I'm fine with another app for production, and I hope my database migrations work...

But I really would like to invite my testers to work on the deployed development version while I continue developing on localhost...

hoping to see an answer here soon...

**Reply by ptcampbell** [reply]
Yes, eventually I'll want a separate production database. But following the [Next.js example using Vercel](https://github.com/vercel/nextjs-subscription-payments) — there is still a distinction between a preview URL and a development URL. Am I just not configuring it properly? I've tried setting a different SITE_URL for `.env.local` and `.env.preview` but it doesn't change the URL generated in magic links, confirmation emails etc.

**Reply by ajoslin103** [reply]
Nice!

I've been using the awesome RedwoodJS for my 1st Supabase project and one of the devs there pointed out the Supabase CLI 

https://community.redwoodjs.com/t/to-localhost-or-not-workaround/2157

The relevant text:

> For Supabase development locally including authentication, have a look at the Supabase CLI
> 
> https://supabase.io/blog/2021/03/31/supabase-cli
> 
> It stands up the entire Supabase stack in Docker via a docker compose file: Kong, Postgres, Postgrest, Real-time, Storage, mailer and Auth.
> 
> You then point your Redwood app in development to the local urls.
>

**Reply by ajoslin103** [reply]
In reply to an email I sent to beta@supabase.io

We don't *officially* support multiple environments yet, but it is possible. 

Here's how the redirects work:

the default redirect is always SITE_URL

there is a `redirectTo` optional parameter in the signIn, signOut, resetPassword js methods: https://supabase.io/docs/reference/javascript/auth-signin#parameters

you can put any value into this redirectTo param but it has to be either:

on the same domain as SITE_URL (e.g. http://local:3000/welcome)

or match *exactly* on of the URLs in the Additional URLs field (including scheme, paths, trailing slashes, and query params)

lmk if something here doesn't make sense,

best,
Ant
CTO @ Supabase

**Reply by Nick-Mazuk** [reply]
My main question is why do the auth urls need to first go to a Supabase owned server? I would have thought that the auth urls would go to your site first when you provide a url. Did start a new discussion because the topic is slightly different: https://github.com/supabase/supabase/discussions/1789

**Reply by aaronlinsky** [reply]
Is there a timeline for multiple environment support for Supabase auth?

**Reply by markitics** [reply]
Thanks for this. Works for me. Took me a while to realise that "A comma separated list of *exact* URLs" means that spaces will mess it up. 
So my list is `https://domain1.com,https://domain2.com,http://local…` etc., 
**not** `https://domain1.com, https://domain2.com, http://local…`

**Comment by denull0** [comment]
Is there a solution to this? It's still not possible to change URLs when using magic links.

**Reply by kiwicopple** [reply]
Yes! The solution is to :

- add all your sites to the "additional site URLs" in the Auth settings
- On each of your sites, create an ENV_VAR with the site url (eg, https://example.com/welcome, https://staging.example.com/welcome)
- Use `redirectTo` param: https://supabase.com/docs/reference/javascript/auth-signin#sign-in-with-redirect

For example:

PROD: 

```js
const { user, session, error } = await supabase.auth.signIn({
  provider: 'github'
}, {
  redirectTo: 'https://example.com/welcome'
})
```

STAGING:


```js
const { user, session, error } = await supabase.auth.signIn({
  provider: 'github'
}, {
  redirectTo: 'https://staging.example.com/welcome'
})
```

**Reply by aaronlinsky** [reply]
Thanks! Does this work with localhost:3000 too?

**Reply by bradledford** [reply]
@aaronlinsky here to say yes, it does work for localhost:3000 as well. Just note the exactness of the URL. They. Must. Match.

**Reply by wiesson** [reply]
> Yes! The solution is to :
> 
> * add all your sites to the "additional site URLs" in the Auth settings
> * On each of your sites, create an ENV_VAR with the site url (eg, https://example.com/welcome, https://staging.example.com/welcome)
> * Use `redirectTo` param: https://supabase.com/docs/reference/javascript/auth-signin#sign-in-with-redirect
> 
> For example:
> 
> PROD:
> 
> ```js
> const { user, session, error } = await supabase.auth.signIn({
>   provider: 'github'
> }, {
>   redirectTo: 'https://example.com/welcome'
> })
> ```
> 
> STAGING:
> 
> ```js
> const { user, session, error } = await supabase.auth.signIn({
>   provider: 'github'
> }, {
>   redirectTo: 'https://staging.example.com/welcome'
> })
> ```

This should be added to the docs!

**Reply by junwang4** [reply]
How exactly do you create an ENV_VAR with the site url  on each of your sites (for example, I have two sites: http://example.com and http://localhost:8100)? How does this work with magic link?

**Reply by t6adev** [reply]
@junwang4 
It's just frontend development topic. Please refer to like this, https://vitejs.dev/guide/env-and-mode.html#env-variables

**Reply by t6adev** [reply]
I tried to apply localhost:3000 and a deployed URL then it was succeeded for the magic link.
My advice is that you please be careful to type those paths correctly. `http://localhost:3000` does not match `http://localhost:3000/` 🙈 

![image](https://user-images.githubusercontent.com/6883692/174606976-093d145a-7eca-46a1-8372-9eeea0cf05ce.png)

Anyway, Thank you all for your advice.

**Reply by callmejumeh** [reply]
@t6adev lining up Site URLs like you've done seems not to work anymore, did you find another work around?
@wiesson Thanks for the breakdown man, but  the "lining up Site URLs" part seems to not work anymore..

Any work around?

**Reply by kevinmlong** [reply]
I'm seeing the same now. It just cobbles the multiple urls together.  @kiwicopple - did something change here with the addition of custom URLs?

**Reply by kevinmlong** [reply]
@callmejumeh - have you figured anything out for addressing this issue?

**Reply by callmejumeh** [reply]
> @callmejumeh - have you figured anything out for addressing this issue?

I managed to find a work around.

SITE_URL in supabase is set to: app.myapp.com
SITE_URL=http://localhost:3000 in my .env.local file.
SITE_URL=app.myapp.com in my Vercel build. (You can set ENV variables in Vercel)

I noticed that process.env.SITE_URL returns undefined if called in client side.
So, I pass process.env.SITE_URL from my getServerSideProps to my component and use that same SITE_URL for the redirectTo param.

`const {data, error} = await supabase.auth.signInWith0Auth({
 provider: 'google', 
 options: {
  redirectTo: siteURL
 } 
}) `

I don't know if it's the best way to do it, but it works.

**Reply by kevinmlong** [reply]
This doesn't seem to work for me. Supabase shows that the site URL is used as an allow-list for redirects. If I try to redirect to a URL that's different than what's in the SITE URL field, it just goes to the SITE_URL.

@kiwicopple - can you provide some guidance here please?

<img width="537" alt="image" src="https://user-images.githubusercontent.com/6371696/221370188-3699031a-08ca-43fc-b0aa-9febf26210b6.png">

**Reply by GaryAustin1** [reply]
The siteURL is really just the default if you don't provide a redirectTo OR the redirectTo URL is not in the Redirect URLs settings under auth.  Must match exactly for the entire URL and there are wildcard options.

**Reply by kevinmlong** [reply]
It is working now. 🤦🏻‍♂️ I was updating code from v1 of the JS library and failed to notice that options object was actually part of the primary parameter and not a second parameter like it used to be I guess.

**Comment by kousun12** [comment]
Is there a good reason not to allow wildcard paths for the allowed redirect urls?

**Reply by isTravis** [reply]
Would also like to discuss this, as I would like to configure the specific entity a redirect will lead to dynamically.

**Comment by J0** [comment]
Hey!
Cross-posting in case anyone stumbles upon this thread:

This should now be supported via wildcard [redirect URIs](https://mobile.twitter.com/supabase/status/1539561445414805506) so you should be able to add the following:

https://mysite.com/users/*

and redirect to https://mysite.com/users/123

Let me know if there are still issues or anything that I've missed.

Thanks!
